### PR TITLE
[17.0][ADD] mail_operating_unit: new module

### PR DESCRIPTION
**Pull request by vvrossem**

### CODE REVIEW

File: mail_operating_unit/models/mail_thread.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+class MailThread(models.AbstractModel):
+    _inherit = "mail.thread"
+
+    def _message_compute_author(
+        self, author_id=None, email_from=None, raise_on_email=True
+    ):
+        alias_domain_id = self.env.context.get("record_alias_domain_id")
+        if alias_domain_id:
+            email_from = (
+                self.env["mail.alias.domain"].browse(alias_domain_id).default_from_email
```
Comment (gurneyalex) [inline mail_operating_unit/models/mail_thread.py]: not super sure this is what we want. 

In my mind, I had more:

```
def _message_author(...):
    author_id, email_from = super()._message_author(...)
    alias_domain_id = self.env.context.get("record_alias_domain_id")
    if alias_domain_id:
        email_from = email_from.split('@')[0] + alias_domain_id.name
    return author_id, email_from 

```

Do you have an opinion on this?

File: mail_operating_unit/models/models.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html).
+
+from odoo import api, models
+
+
+class BaseModel(models.AbstractModel):
+    _inherit = "base"
+
+    @api.model
+    def _mail_get_operating_unit(self):
```
Comment (gurneyalex) [inline mail_operating_unit/models/models.py:10]: add a docstring

File: mail_operating_unit/models/models.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            return operating_unit
+
+        user = self.env.user
+        user_operating_units = user.operating_unit_ids
+        if user_operating_units:
+            if len(user_operating_units) == 1:
+                return user_operating_units
+            unique_alias_domains = user_operating_units.mapped("alias_domain_id.id")
+            if len(user_operating_units) > 1 and len(unique_alias_domains) == 1:
+                return user_operating_units[0]
```
Comment (gurneyalex) [inline mail_operating_unit/models/models.py]: we want the first user_operating_unit which has an alias_domain_id, which can be the second in the list.

File: mail_operating_unit/models/models.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    @api.model
+    def _mail_get_operating_unit(self):
+        return (
+            self.operating_unit_id
+            if "operating_unit_id" in self and self.operating_unit_id
+            else False
+        )
+
+    def _mail_get_operating_units(self):
```
Comment (gurneyalex) [inline mail_operating_unit/models/models.py:24]: docstring?

**Review by yankinmax — APPROVED** [review]

File: mail_operating_unit/models/mail_template.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+class MailTemplate(models.Model):
+    _inherit = "mail.template"
+
+    operating_unit_id = fields.Many2one("operating.unit", string="Operating Unit")
+
+    def send_mail(
+        self, res_id, force_send=False, raise_exception=False, email_values=None
+    ):
+        """Include in the email values the alias domain ID of the current OU if any"""
```
Comment (Camille0907) [inline mail_operating_unit/models/mail_template.py]:
```suggestion
        # Include in the email values the alias domain ID of the current OU if any
```
Using `"""` will erase the previous doctring of the method with yours, I don't think this is what we want :slightly_smiling_face:

File: mail_operating_unit/models/models.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    def _mail_get_alias_domains(self, default_company=False):
+        """Use operating unit's alias domain, if any.
+
+        If an OU is associated with a record, its alias domain is used.
+        Otherwise, the default alias domain is applied.
+        """
+        alias_domains = super()._mail_get_alias_domains(default_company=default_company)
+        return {
+            record.id: (
+                record._mail_get_operating_units().alias_domain_id
```
Comment (Camille0907) [inline mail_operating_unit/models/models.py]: Maybe we could store  `record._mail_get_operating_units()` in a variable ?

File: mail_operating_unit/tests/test_fake_partner_mail_operating_unit.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        )
+        self.operating_unit = self.env["operating.unit"].create(
+            {
+                "name": "Operating Unit",
+                "code": "OU",
+                "company_id": self.company.id,
+                "partner_id": self.company.partner_id.id,
+                "alias_domain_id": self.operating_unit_alias_domain.id,
+            }
+        )
```
Comment (Camille0907) [inline mail_operating_unit/tests/test_fake_partner_mail_operating_unit.py]: This should go in setUpClass, no ?

**Review by Camille0907 — APPROVED** [review]
LGTM, nitpicking on a few points only :slightly_smiling_face:

**Comment by yankinmax** [comment]
@vvrossem can you pls address @Camille0907 suggestions?

File: mail_operating_unit/models/models.py (hunk 6 above)
Comment (vvrossem) [reply mail_operating_unit/models/models.py]: here comes the walrus operator! :rocket:

Comment (yankinmax) [reply mail_operating_unit/models/models.py]: Perfect choice! Thanks

**Comment by yankinmax** [comment]
@AaronHForgeFlow can you pls take a look here and push merge?

File: mail_operating_unit/models/mail_template.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+    def send_mail(
+        self, res_id, force_send=False, raise_exception=False, email_values=None
+    ):
+        # Include in the email values the alias domain ID of the current OU if any
+        if self.operating_unit_id and self.operating_unit_id.alias_domain_id:
+            email_values = dict(
+                email_values or {},
+                record_alias_domain_id=self.operating_unit_id.alias_domain_id.id,
+            )
```
Comment (simahawk) [inline mail_operating_unit/models/mail_template.py]:
```suggestion
        email_values = email_values or {}
        if self.operating_unit_id and self.operating_unit_id.alias_domain_id:
            email_values["record_alias_domain_id"] = self.operating_unit_id.alias_domain_id.id,
```

File: mail_operating_unit/tests/__init__.py [hunk 9]
```diff
@@ -0,0 +1,3 @@
+from . import common
```
Comment (simahawk) [inline mail_operating_unit/tests/__init__.py]: no need to import common

File: mail_operating_unit/tests/test_fake_partner_mail_operating_unit.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
+from odoo_test_helper import FakeModelLoader
+
+from odoo.tests import common
+
+
+class MailOperatingUnitCommon(common.TransactionCase):
```
Comment (simahawk) [inline mail_operating_unit/tests/test_fake_partner_mail_operating_unit.py]: pls use BaseCommon as base class to get tracking disabled

File: mail_operating_unit/tests/test_fake_partner_mail_operating_unit.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
+from odoo_test_helper import FakeModelLoader
```
Comment (simahawk) [inline mail_operating_unit/tests/test_fake_partner_mail_operating_unit.py:4]: header missing

File: mail_operating_unit/tests/common.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
+from odoo.models import Command
```
Comment (simahawk) [inline mail_operating_unit/tests/common.py:4]: header missing

File: mail_operating_unit/tests/test_mail_operating_unit.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
+from odoo.addons.mail_operating_unit.tests.common import MailOperatingUnitCommon
```
Comment (simahawk) [inline mail_operating_unit/tests/test_mail_operating_unit.py]: header missing

Comment (simahawk) [reply mail_operating_unit/tests/test_mail_operating_unit.py]: you should import from `.common`

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by vvrossem** [comment]
thank you for reviewing @simahawk :orange_heart:

**Review by simahawk — APPROVED** [review]

**Pull request merged**